lib/org/chromium/webidl/IDLFragmentExtractor.js

```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  package: 'org.chromium.webidl',
+  name: 'IDLFragmentExtractor',
+  documentation: 'extracts IDL Fragments from HTML files',
+  requires: [
+    'foam.parsers.html.HTMLLexer'
+  ],
+
+  properties: [
+    {
+      name: 'file',
+      of: 'org.chromium.webidl.HTMLFileContent',
+      postSet: function(_, file) {
+        this.idlFragments = this.extract();
+      },
+    },
+    {
+      name: 'idlFragments',
+      // description: 'all idlFragments found in the parsed HTML file',
+    },
+  ],
+
+  methods: [
+    function extract() {
+      var self = this;
+      var lexer = self.HTMLLexer.create();
+      var OPEN = lexer.TagType.OPEN.name;
+      var CLOSE = lexer.TagType.CLOSE.name;
+      var extractAttr = function(node, attrName) {
+        var retVal;
+        node.attributes.some(function(attr) {
+          if (attr.name === attrName) {
+            retVal = attr.value.split(' ');
+          }
+        });
+        return retVal;
+      };
+
+      var results = lexer.parseString(self.file.content).value;
+
+      var idlFragments = [];
+      var tagStack = [];            // Used for tag matching.
+      var tagMatching = true;       // Set whe not inside a pre tag of interest.
+      var exclude = false;          // Set when inside example or note class.
+      var content = '';             // Used to group together related content.
+      for (var i = 0; i < results.length; i++) {
+        var item = results[i];
+        var isTag = lexer.Tag.isInstance(item);
+
+        if (!tagMatching) {
+          // Ignoring all tags. Only extracting text within pre tags.
+          if (isTag && item.nodeName === 'pre') {
+            tagMatching = true;
+            tagStack.pop();
+            idlFragments.push(lexer.lib.unescapeString(content));
+            content = '';
+          } else {
+            content += isTag ? '' : item;
+          }
+        } else if (isTag) {
+          var top = tagStack[tagStack.length - 1];
+          var classes = extractAttr(item, 'class');
+          var isIDL = classes && classes.includes('idl');
+          var isExcluded = function(cls) {
+            return cls && (cls.includes('example') || cls.includes('note'));
+          };
+
+          if (item.type.name === OPEN) {
+            if (isExcluded(classes)) {
+              exclude = true;
+            } else if (!exclude && item.nodeName === 'pre' && isIDL) {
+              // Found a <pre class="idl.*">.
+              // Ignore tags and only extract text now.
+              tagMatching = false;
+            }
+            tagStack.push(item);
+          } else if (top && item.type.name === CLOSE && top.nodeName === item.nodeName) {
+            var parentCls = extractAttr(top, 'class');
+            if (isExcluded(parentCls)) exclude = false;
```

Comment:
I have not yet observed an instance of an exclude tag nested within another, but there is a relatively simple fix to this issue (using a stack to track excluded tags instead of a bool), so that has been implemented. Will be in next set of changes.